### PR TITLE
chore(web): eslint react hook rules (part 1)

### PR DESCRIPTION
**Pull request by LetItRock**
### What change does this PR introduce?

Enable ESLinkt React hook rules: https://legacy.reactjs.org/docs/hooks-rules.html
The plan is to fix all the warnings left ~80 in a smaller PRs.
- fixed all errors
- fixed a few warnings
- refactored the spotlight context, it's working much better now and a couple of re-renders are saved

### Other information (Screenshots)

https://github.com/novuhq/novu/assets/2607232/fe4deea0-d02c-4825-96d5-ae2751aa7edb

### CODE REVIEW

File: apps/web/.eslintrc.js [hunk 1]
```diff
@@ -35,13 +35,15 @@ module.exports = {
         format: ['PascalCase', 'camelCase', 'UPPER_CASE'],
       },
     ],
+    "react-hooks/rules-of-hooks": "error",
+    "react-hooks/exhaustive-deps": "warn"
```
Comment (LetItRock) [inline apps/web/.eslintrc.js:39]: configuration as they suggest in the docs: https://www.npmjs.com/package/eslint-plugin-react-hooks

File: apps/web/src/components/layout/AppLayout.tsx [hunk 2]
```diff
@@ -60,17 +60,9 @@ export function AppLayout() {
...
   );
 }
-
-function SupportChatProvider({ children }) {
-  if (INTERCOM_APP_ID) {
-    return <IntercomProvider appId={INTERCOM_APP_ID}>{children}</IntercomProvider>;
-  }
-
-  return children;
-}
```
Comment (LetItRock) [inline apps/web/src/components/layout/AppLayout.tsx:76]: relates to conditionally calling hook, the code change in a file below...
when there is no `INTERCOM_APP_ID` the library doesn't show any warnings until it's initialized... but we don't do that if we don't have the `INTERCOM_APP_ID` value

File: apps/web/src/components/providers/SpotlightProvider.tsx [hunk 3]
```diff
@@ -22,20 +22,21 @@ export const useSpotlightContext = (): ISpotlightContext => useContext(Spotlight
...
+    [setItems]
+  );
 
-  const removeItem = (id: string) => {
-    setItems(items.filter((item) => item.id !== id));
-  };
+  const contextValue = useMemo(() => ({ items, addItem, removeItems }), [items, addItem, removeItems]);
 
-  return <SpotlightContext.Provider value={{ items, addItem, removeItem }}>{children}</SpotlightContext.Provider>;
+  return <SpotlightContext.Provider value={contextValue}>{children}</SpotlightContext.Provider>;
```
Comment (LetItRock) [inline apps/web/src/components/providers/SpotlightProvider.tsx:41]: the value should be memoized between re-renders

File: apps/web/src/components/layout/components/HeaderNav.tsx [hunk 4]
```diff
@@ -177,7 +179,7 @@ export function HeaderNav({}: Props) {
         </Link>
         <Group>
           <ActionIcon variant="transparent" onClick={() => toggleColorScheme()}>
-            <Tooltip label={themeTitle()}>
+            <Tooltip label={themeTitle}>
               <div>{Icon()}</div>
```
Comment (ainouzgali) [inline apps/web/src/components/layout/components/HeaderNav.tsx:183]: Should Icon() here also be changed to <Icon /> ?

**Review by ainouzgali — APPROVED** [review]

**Pull request merged**